### PR TITLE
add support for nested outlines

### PR DESCRIPTION
**Pull request by kaymmm**
I added new mappings for C-t, C-d, >>, and << to promote and demote bullets. There is also a new option containing an ordered list of nested outline levels: `g:bullets_nested_outline_levels`.

I added specs for testing, along with a new spec for alphabetic lists (maybe should be a separate pull). The new alphabetic list currently fails because it doesn't correctly identify wrapped lines that 'look' like alphabetic bullets. Also still need additional tests for nested bullets to test for edge cases, but I wanted to get this pull request in for now.

### CODE REVIEW

**Comment by kaymmm** [comment]
Didn't previously see #33, but this PR addresses that issue. 

I might also look into adding an option to satisfy, at least partially, #46. However, there are a bunch of other CommonMark specs that might not be satisfied with the current implementation (I haven't fully reviewed all of them). This might also raise an issue with tabs versus spaces.

**Comment by dkarter** [comment]
Thanks for the PR @kaymmm! I'll take a look tomorrow!

File: plugin/bullets.vim [hunk 1]
```diff
@@ -573,6 +596,113 @@ endfun
...
+          if g:bullets_delete_last_bullet_if_empty
+            let l:orig_line = s:parse_bullet(l:lnum, getline(l:lnum))
+            if l:orig_line != []
+              call setline(l:lnum, l:orig_line[0].leading_space . l:orig_line[0].text_after_bullet)
+            endif
+          endif
+        endif
+      endif
+    endif
+  endif
```
Comment (dkarter) [inline plugin/bullets.vim]: I'm a little concerned by all the `if` statements. Maybe there's a way to break some stuff into smaller private functions? Perhaps annotating with comments above each case what it is supposed to handle would help? (there's already some comments which I find very helpful but there are a lot of cases that don't)

File: plugin/bullets.vim [hunk 2]
```diff
@@ -573,6 +596,113 @@ endfun
...
+endfun
+
+fun! s:bullet_demote()
+  call s:change_bullet_level(-1)
+endfun
+fun! s:bullet_promote()
```
Comment (dkarter) [inline plugin/bullets.vim]: can we add some space here?

File: plugin/bullets.vim [hunk 3]
```diff
@@ -573,6 +596,113 @@ endfun
...
+                \ l:curr_bullet.leading_space
+                \ . l:curr_bullet.text_after_bullet
+        endif
+
+        if l:next_bullet_str !=# ''
+          call setline(l:lnum, l:next_bullet_str)
+          execute "normal! $" 
+
+        else
+          if g:bullets_delete_last_bullet_if_empty
```
Comment (dkarter) [inline plugin/bullets.vim]: This can just be `elseif`, I think..

File: plugin/bullets.vim [hunk 4]
```diff
@@ -55,6 +55,13 @@ while s:power >= 0
   let s:abc_max += pow(26,s:power)
   let s:power -= 1
 endwhile
+
+if !exists('g:bullets_outline_levels')
+  " Capitalization matters: all caps will make the symbol caps, lower = lower
+  " Standard bullets should include the marker symbol after 'std'
+  let g:bullets_outline_levels = ['ROM', 'ABC', 'num', 'abc', 'rom', 'std-', 'std*', 'std+']
+endif
```
Comment (dkarter) [inline plugin/bullets.vim]: I love this!

File: plugin/bullets.vim [hunk 5]
```diff
@@ -573,6 +596,113 @@ endfun
...
+        else
+          " We're outside of the defined outline levels
+          let l:next_bullet_str =
+                \ l:curr_bullet.leading_space
+                \ . l:curr_bullet.text_after_bullet
+        endif
+
+        if l:next_bullet_str !=# ''
+          call setline(l:lnum, l:next_bullet_str)
+          execute "normal! $"
```
Comment (dkarter) [inline plugin/bullets.vim]: Minor linting: Vint wants this to be single quoted string

File: spec/alphabetic_bullets_spec.rb [hunk 6]
```diff
@@ -132,6 +132,33 @@
...
+    #     a. first bullet might not catch
+    #     	me. second line.
+    #   INIT
+    #     # Hello there
+    #     a. first bullet might not catch
+    #     	me. second line.
+    #     not a bullet
+    #   EXPECTED
+    # end
+
```
Comment (dkarter) [inline spec/alphabetic_bullets_spec.rb]: That's a bummer, I think I have a solution though, for alphabetic bullets, if nested we can check that the first one is `a.` or `A.` - not necessary to fix this in this PR but good to keep this here as a TODO.

File: plugin/bullets.vim (hunk 1 above)
Comment (kaymmm) [reply plugin/bullets.vim]: Yeah, I was looking for a cleaner way of handling this mess when I was writing it. For clarity's sake, I moved some of those checks into empty returns so that everything below it is more readable. I can offload some of this into other functions, but they would be single-use functions that require a bunch of parameters and this doesn't seem like the cleanest solution either.

File: spec/alphabetic_bullets_spec.rb (hunk 6 above)
Comment (kaymmm) [reply spec/alphabetic_bullets_spec.rb]: I had a similar thought about checking to see if an item is the first in the list, and if not, check to see if the previous sibling bullet is one increment lower. This solution breaks if the bullets are not sequential because of reordering, inserting new bullets above the current one, etc., but might be better than nothing.

Another option would be to follow the Pandoc-Markdown rules for [ordered lists](https://pandoc.org/MANUAL.html#ordered-lists), which only allow single letter bullets (except roman numerals) and require two spaces after capital letters to disambiguate from initialized names (e.g., `G. Washington` vs. `G.  Washington`).

**Comment by dkarter** [comment]
Thanks so much for this PR! Awesome work! ✨
